src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,493 @@
+		$successful = array();
+		$failed     = array();
+
+		/*
+		 * Using `get_plugin_data()` instead has produced warnings/errors
+		 * as the files may not be in place at this time.
+		 */
+		$plugins = get_plugins();
+
+		foreach ( static::$current_plugins->response as $k => $update ) {
+			$item = static::$current_plugins->response[ $k ];
+			$name = $plugins[ $update->plugin ]['Name'];
+
+			/*
+			 * This appears to be the only way to get a plugin's older version
+			 * at this stage of an auto-update when not implementing this
+			 * feature directly in Core.
+			 */
+			$current_version = static::$current_plugins->checked[ $update->plugin ];
+
+			/*
+			 * The `current_version` property does not exist yet. Add it.
+			 *
+			 * `static::$current_plugins->response[ $k ]` is an instance of `stdClass`,
+			 * so this should not fall victim to PHP 8.2's deprecation of
+			 * dynamic properties.
+			 */
+			$item->current_version = $current_version;
+
+			$plugin_result = (object) array(
+				'name' => $name,
+				'item' => $item,
+			);
+
+			$success = array_diff( self::$processed, self::$fatals );
+
+			if ( in_array( $update->plugin, $success, true ) ) {
+				$successful['plugin'][] = $plugin_result;
+				continue;
+			}
+
+			if ( in_array( $update->plugin, self::$fatals, true ) ) {
+				$failed['plugin'][] = $plugin_result;
+			}
+		}
+
+		$automatic_upgrader      = new WP_Automatic_Updater();
+		$send_plugin_theme_email = new ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
+		$send_plugin_theme_email->setAccessible( true );
+		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
+
+		remove_filter( 'auto_plugin_theme_update_email', array( $this, 'auto_update_rollback_message' ), 10 );
+	}
+
+	/**
+	 * Add auto-update failure message to email.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param array  $email {
+	 *     Array of email arguments that will be passed to wp_mail().
+	 *
+	 *     @type string $to      The email recipient. An array of emails
+	 *                           can be returned, as handled by wp_mail().
+	 *     @type string $subject The email's subject.
+	 *     @type string $body    The email message body.
+	 *     @type string $headers Any email headers, defaults to no headers.
+	 * }
+	 * @param string $type               The type of email being sent. Can be one of 'success', 'fail', 'mixed'.
+	 * @param array  $successful_updates A list of updates that succeeded.
+	 * @param array  $failed_updates     A list of updates that failed.
+	 *
+	 * @return array
+	 */
+	public function auto_update_rollback_message( $email, $type, $successful_updates, $failed_updates ) {
+		if ( empty( $failed_updates ) ) {
+			return $email;
+		}
+		$body   = explode( "\n", $email['body'] );
+		$failed = __( 'These plugins failed to update or may have been rolled back due to detection of a fatal error:' );
```

Comment:
Is ”rolled back” a understandable term to show a user? It is really hard to translate in a meaningfull way. Can the sentence be:
`These plugins failed to update or may have been retored due to detection of a fatal error:` ?

Reply:
Or even:
`These plugins failed to update or may have been restored to their previous version due to detection of a fatal error:`

Reply:
This string was the thing that got me going, so yeah, await a decision there and maybe apply changes here post decision.

Reply:
"Returned" or "not changed"... That's what I'd use.